blueos_repository/consolidate.py

```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/env python3
+import asyncio
+import dataclasses
+import json
+from pathlib import Path
+from typing import Any, AsyncIterable, Dict, List, Optional
+
+import aiohttp
+from registry import Registry
+
+REPO_ROOT = "https://raw.githubusercontent.com/bluerobotics/BlueOS-Extensions-Repository/master/repos"
+
+
+class EnhancedJSONEncoder(json.JSONEncoder):
+    def default(self, o):  # type: ignore
+        if dataclasses.is_dataclass(o):
+            return dataclasses.asdict(o)
+        return super().default(o)
+
+
+@dataclasses.dataclass
+class Author:
+    name: str
+    email: str
+
+    @staticmethod
+    def from_json(json_dict: Dict[str, str]) -> "Author":
+        return Author(name=json_dict["name"], email=json_dict["email"])
+
+
+@dataclasses.dataclass
+class Company:
+    name: str
+    about: Optional[str]
+    email: Optional[str]
+
+    @staticmethod
+    def from_json(json_dict: Dict[str, str]) -> "Company":
+        return Company(name=json_dict["name"], email=json_dict.get("email", None), about=json_dict.get("about", None))
+
+
+# pylint: disable=too-many-instance-attributes
+@dataclasses.dataclass
+class Version:
+    permissions: Optional[Dict[str, Any]]
+    requirements: Optional[str]
+    tag: Optional[str]
+    website: str
+    authors: List[Author]
+    docs: Optional[str]
+    readme: Optional[str]
+    company: Optional[Company]
+    support: Optional[str]
+
+
+@dataclasses.dataclass
+class RepositoryEntry:
+    identifier: str
+    name: str
+    description: str
+    docker: str
+    versions: Dict[str, Version]
+    extension_logo: Optional[str]
+    company_logo: Optional[str]
+
+
+class Consolidator:
+    registry = Registry()
+    consolidated_data: List[RepositoryEntry] = []
+
+    @staticmethod
+    async def fetch_readme(url: str) -> str:
+        if not url.startswith("http"):
+            print(f"Invalid Readme url: {url}")
+            return ""
+        async with aiohttp.ClientSession() as session:
+            async with session.get(url) as resp:
+                if resp.status != 200:
+                    print(f"Error status {resp.status}")
+                    raise Exception(f"Could not get readme {url}")
+                return await resp.text()
+
+    async def all_repositories(self) -> AsyncIterable[RepositoryEntry]:
+        repos = Path("./repos")
+        for repo in repos.glob("**/*.json"):
```

Comment:
`metadata` could be a better name to describe the json file

Reply:
done